tests/unit/test_register_user_action.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Register user action unit tests."""
+
+# pylint: disable=protected-access
+
+import typing
+import unittest.mock
+
+import ops
+import pytest
+from ops.charm import ActionEvent
+from ops.testing import Harness
+
+import synapse
+from user import User
+
+
+@pytest.mark.parametrize("harness", [0], indirect=True)
+def test_register_user_action(
+    harness_server_name_configured: Harness, monkeypatch: pytest.MonkeyPatch
+) -> None:
+    """
+    arrange: start the Synapse charm, set Synapse container to be ready and set server_name.
+    act: run register-user action.
+    assert: Synapse charm should reset the instance.
+    """
+    harness = harness_server_name_configured
+    harness.set_leader(True)
+    get_registration_mock = unittest.mock.Mock(return_value="shared_secret")
+    monkeypatch.setattr("synapse.get_registration_shared_secret", get_registration_mock)
+    register_user_mock = unittest.mock.MagicMock()
+    monkeypatch.setattr("synapse.register_user", register_user_mock)
+    user = "username"
+    event = unittest.mock.MagicMock(spec=ActionEvent)
+    event.params = {
+        "username": user,
+        "admin": "no",
+    }
+    # Calling to test the action since is not possible calling via harness
+    harness.charm._on_register_user_action(event)
+    assert event.set_results.call_count == 1
+    event.set_results.assert_called_with(
+        {"register-user": True, "user-password": unittest.mock.ANY}
+    )
+    assert isinstance(harness.model.unit.status, ops.ActiveStatus)
+
+
+@pytest.mark.parametrize("harness", [0], indirect=True)
+def test_register_user_registration_none(
+    harness_server_name_configured: Harness, monkeypatch: pytest.MonkeyPatch
+) -> None:
+    """
+    arrange: start the Synapse charm, set Synapse container to be ready and set server_name.
+    act: run register-user action.
+    assert: event fails if registration shared secret is not found.
+    """
+    harness = harness_server_name_configured
+    harness.set_leader(True)
```

Comment:
Would it be worth doing this in the `harness_server_name_configured` fixture?

Reply:
Yes, changed.